File: packages/go-server/dynamic_config_test.go
package savvagent

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestDynamicConfiguration(t *testing.T) {
	t.Run("Evaluate returns configuration when flag is enabled", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"primaryColor": "#007bff",
				"fontSize":     float64(16),
			},
			"limits": map[string]interface{}{
				"maxItems": float64(100),
			},
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
				"flag_id":       "flag-123",
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		result, err := client.Evaluate("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if !result.Value {
			t.Error("Expected flag to be enabled")
		}
		if result.Configuration == nil {
			t.Error("Expected configuration to be present")
		}
		if result.Configuration["theme"] == nil {
			t.Error("Expected theme in configuration")
		}
	})

	t.Run("GetConfig returns configuration when flag is enabled", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"apiEndpoint": "https://api.example.com",
			"timeout":     float64(5000),
			"retries":     float64(3),
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		config, err := client.GetConfig("api-settings", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if config == nil {
			t.Fatal("Expected configuration to be present")
		}
		if config["apiEndpoint"] != "https://api.example.com" {
			t.Error("Configuration value mismatch")
		}
	})

	t.Run("GetConfig returns nil when flag is disabled", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value": false,
				"configuration": map[string]interface{}{
					"some": "config",
				},
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		config, err := client.GetConfig("api-settings", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if config != nil {
			t.Error("Expected nil configuration when flag is disabled")
		}
	})

	t.Run("GetVariation returns variation details", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"algorithm": "ml_v2",
			"weight":    float64(2.0),
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"variation":     "variant_b",
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		result, err := client.GetVariation("search-algorithm", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if result.Variation != "variant_b" {
			t.Errorf("Expected variation 'variant_b', got '%s'", result.Variation)
		}
		if !result.Enabled {
			t.Error("Expected flag to be enabled")
		}
		if result.Configuration == nil {
			t.Error("Expected configuration to be present")
		}
	})

	t.Run("GetVariation defaults to 'control' when not specified", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value": true,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		result, err := client.GetVariation("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if result.Variation != "control" {
			t.Errorf("Expected default variation 'control', got '%s'", result.Variation)
		}
	})

	t.Run("Configuration is cached along with flag value", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"setting": "value",
		}

		callCount := 0
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			callCount++
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:   "sdk_test_key",
			BaseURL:  server.URL,
			CacheTTL: 5 * time.Second,
		})

		// First call
		result1, _ := client.Evaluate("test-flag", nil)
		if result1.Reason != "evaluated" {
			t.Error("First call should be evaluated")
		}

		// Second call should be cached
		result2, _ := client.Evaluate("test-flag", nil)
		if result2.Reason != "cached" {
			t.Error("Second call should be cached")
		}

		if callCount != 1 {
			t.Errorf("Expected 1 API call, got %d", callCount)
		}

		// Both should have configuration
		if result1.Configuration == nil || result2.Configuration == nil {
			t.Error("Both results should have configuration")
		}
	})

	t.Run("Backward compatibility with IsEnabled", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value": true,
				"configuration": map[string]interface{}{
					"some": "config",
				},
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		enabled := client.IsEnabled("test-flag", nil)

		if !enabled {
			t.Error("Expected flag to be enabled")
		}
	})

	t.Run("Works with API responses without configuration", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value": true,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		result, err := client.Evaluate("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if !result.Value {
			t.Error("Expected flag to be enabled")
		}
		// Configuration should be nil or empty
	})
}

File: packages/node-server/src/dynamic-config.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { FlagClient } from './client';
import { FlagClientConfig } from './types';

// Mock fetch globally
global.fetch = vi.fn();

describe('Dynamic Configuration Tests', () => {
  let client: FlagClient;
  let mockFetch: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    mockFetch = global.fetch as ReturnType<typeof vi.fn>;
    mockFetch.mockClear();

    const config: FlagClientConfig = {
      apiKey: 'sdk_test_key',
      applicationId: 'test-app',
      baseUrl: 'https://api.test.com',
      enableRealtime: false,
      enableTelemetry: false,
    };

    client = new FlagClient(config);
  });

  describe('evaluate() with configuration', () => {
    it('should return configuration when flag is enabled', async () => {
      const mockConfig = {
        theme: { primaryColor: '#007bff', fontSize: 16 },
        limits: { maxItems: 100 },
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          configuration: mockConfig,
          flagId: 'flag-123',
        }),
      });

      const result = await client.evaluate('test-flag');

      expect(result.value).toBe(true);
      expect(result.configuration).toEqual(mockConfig);
      expect(result.reason).toBe('evaluated');
    });

    it('should return null configuration when not provided by API', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          flagId: 'flag-123',
        }),
      });

      const result = await client.evaluate('test-flag');

      expect(result.value).toBe(true);
      expect(result.configuration).toBeUndefined();
    });

    it('should cache configuration along with flag value', async () => {
      const mockConfig = { setting: 'value' };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          configuration: mockConfig,
        }),
      });

      // First call - fetches from API
      const result1 = await client.evaluate('test-flag');
      expect(result1.configuration).toEqual(mockConfig);
      expect(result1.reason).toBe('evaluated');

      // Second call - returns from cache
      const result2 = await client.evaluate('test-flag');
      expect(result2.configuration).toEqual(mockConfig);
      expect(result2.reason).toBe('cached');
      expect(mockFetch).toHaveBeenCalledTimes(1);
    });
  });

  describe('getConfig()', () => {
    it('should return configuration when flag is enabled', async () => {
      const mockConfig = {
        apiEndpoint: 'https://api.example.com',
        timeout: 5000,
        retries: 3,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          configuration: mockConfig,
        }),
      });

      const config = await client.getConfig('api-settings');

      expect(config).toEqual(mockConfig);
    });

    it('should return null when flag is disabled', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: false,
          configuration: { some: 'config' },
        }),
      });

      const config = await client.getConfig('api-settings');

      expect(config).toBeNull();
    });

    it('should return default value when flag is disabled', async () => {
      const defaultConfig = { fallback: true };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ value: false }),
      });

      const config = await client.getConfig('api-settings', undefined, defaultConfig);

      expect(config).toEqual(defaultConfig);
    });

    it('should return default value when configuration is missing', async () => {
      const defaultConfig = { default: 'value' };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ value: true }),
      });

      const config = await client.getConfig('api-settings', undefined, defaultConfig);

      expect(config).toEqual(defaultConfig);
    });
  });

  describe('getVariation()', () => {
    it('should return variation details with configuration', async () => {
      const mockConfig = {
        algorithm: 'ml_v2',
        weight: 2.0,
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          variation: 'variant_b',
          configuration: mockConfig,
        }),
      });

      const result = await client.getVariation('search-algorithm');

      expect(result.variation).toBe('variant_b');
      expect(result.enabled).toBe(true);
      expect(result.configuration).toEqual(mockConfig);
    });

    it('should default to "control" when variation is not specified', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ value: true }),
      });

      const result = await client.getVariation('test-flag');

      expect(result.variation).toBe('control');
      expect(result.enabled).toBe(true);
    });

    it('should include disabled state in variation result', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: false,
          variation: 'control',
        }),
      });

      const result = await client.getVariation('test-flag');

      expect(result.enabled).toBe(false);
      expect(result.variation).toBe('control');
    });
  });

  describe('Backward Compatibility', () => {
    it('should maintain isEnabled() behavior', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          value: true,
          configuration: { some: 'config' },
        }),
      });

      const enabled = await client.isEnabled('test-flag');

      expect(enabled).toBe(true);
    });

    it('should work with API responses without configuration field', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ value: true }),
      });

      const result = await client.evaluate('test-flag');

      expect(result.value).toBe(true);
      expect(result.configuration).toBeUndefined();
    });
  });

  describe('Error Handling', () => {
    it('should handle API errors gracefully in getConfig', async () => {
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 500,
        statusText: 'Internal Server Error',
      });

      const config = await client.getConfig('test-flag');

      expect(config).toBeNull();
    });

    it('should return default value on network error', async () => {
      const defaultConfig = { fallback: true };

      mockFetch.mockRejectedValueOnce(new Error('Network error'));

      const config = await client.getConfig('test-flag', undefined, defaultConfig);

      expect(config).toEqual(defaultConfig);
    });
  });
});
